tdrs-backend/tdpservice/settings/common.py

```diff
@@ -10,6 +10,58 @@
 
 BASE_DIR = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))
 
+# NOTE: These may be overridden by system or env_file variables
```

Comment:
I think it's great to define some defaults that can be used and overridden on different environments, but I'm worried about introducing another means of setting env vars. Right now we are using a combinations of env vars set through `.env` files, defined directly in the docker-compose.yml file, and within CircleCI. This would introduce another convention, and may not be immediately apparent to a new dev where these values are being set.

Would it be possible to accomplish this same end result though the use of an `.env` file since that is the convention we are using for local development?

Reply:
If you look at the original issue and AC's I think this pretty much would undo what the entire issue was meant to address, no? https://github.com/raft-tech/TANF-app/issues/510

Reply:
It might also be useful to note that this doesn't really set the environment variables, but provides defaults if they are missing. Environment variables themselves will still be set in the `.env`

Reply:
Taking a step back, I think the main goal of 510 is to come up with a more coherent solution to handling env vars locally. Putting myself in the shoes of a new dev onboarding to the project, I'm not sure I would know that missing env vars would be falling back to those defined in `common.py` (at least without additional documentation explaining the order of inheritance of env vars).

I'm not personally reading the issue as do not use .env files, but rather addressing the current local development env configuration where env vars have to be set by running a bash script or other means, and just a general sense of confusion about where vars are defined. Let's sync up next week though, because I'm less aware of the origin story of 510 so there's a good change I'm missing some context so would be worth it to spend a few minutes getting on the same page!

Reply:
@hursey013 I updated the backend README to document the order of inheritance for environment variables for both local and CI environments. Please review and let me know if that addresses everything.

Reply:
Hmmmm, just want to note that these defaults won't be correct for all environments. 

For example, in production, we wouldn't want to use the identity sandbox configuration. 

I do think that backing out these changes would simplify ENV handling, but I could be persuaded either way.

Reply:
The key benefit to doing it this way is having easily *overridable* defaults. We could define these in docker-compose as defaults but when doing so it's not easy to override them during local development without declaring that variable in multiple compose files.

Take this example from the base docker-compose:
```
    environment:
      - DJANGO_SECRET_KEY=local
```
I actually have this variable defined in my `.env` file but if I launch the backend my override is ignored:
```
DJANGO_SECRET_KEY=local-mega-secret  # As set in .env

❯ docker-compose -f docker-compose.yml -f docker-compose.local.yml up -d                                                                                                                                        ─╯
tdrs-backend_postgres_1 is up-to-date
tdrs-backend_zaproxy_1 is up-to-date
Recreating tdrs-backend_web_1 ... done

❯ docker-compose exec web /bin/bash                                                                                                                                                                             ─╯
root@35ece2161394:/tdpapp# echo $DJANGO_SECRET_KEY
local
```
When defaults are set via the `environment` directive in docker-compose the only way to override them is to repeat the environment definition in a separate compose file. Here's an example with it added to the `docker-compose.local.yml`:
```
    environment:
      - DJANGO_SECRET_KEY=overridden
```
```
❯ docker-compose down && docker-compose -f docker-compose.yml -f docker-compose.local.yml up -d 
❯ docker-compose exec web /bin/bash                                                                                                                                                                             ─╯
root@75cbe0429e8d:/tdpapp# 
root@75cbe0429e8d:/tdpapp# echo $DJANGO_SECRET_KEY
overridden
```

With the defaults set using `os.environ.setdefault` we can easily override them in any environment. For example, consider this default:
```
os.environ.setdefault("FRONTEND_BASE_URL", "http://localhost:3000")
```
If I don't set anything for that var in `.env` the host won't have the variable set, but when attempting to reference it in python the default will be applied:
```
root@bf6f151df2cd:/tdpapp# echo $FRONTEND_BASE_URL

root@bf6f151df2cd:/tdpapp# python manage.py shell
In [1]: import os

In [2]: os.environ['FRONTEND_BASE_URL']
Out[2]: 'http://localhost:3000'
```
Now I can easily override that default by supplying the value in `.env`:
```
FRONTEND_BASE_URL=http://zombo.com
```
```
❯ docker-compose down && docker-compose -f docker-compose.yml -f docker-compose.local.yml up -d                                                                                                                 ─╯
❯ docker-compose exec web /bin/bash                                                                                                                                                                             ─╯
root@80473861c2ae:/tdpapp# echo $FRONTEND_BASE_URL
http://zombo.com
root@80473861c2ae:/tdpapp# python manage.py shell
In [1]: import os

In [2]: os.environ['FRONTEND_BASE_URL']
Out[2]: 'http://zombo.com'
```
In production contexts environment variables are managed by Cloud.gov, so they can be set there as needed and the defaults defined in settings will be overridden as expected.

Reply:
Thank you for the thorough explanation! In the future we may want to move some of these defaults into more specific settings files like `test.py`, `local.py`, since I'm seeing that `production.py` imports the `Common` settings -- but this works for me for now, especially since our focus here is on fixing the local dev experience!

Reply:
Definitely, I'd like to expand on this further in the future. Ideally within Django we'd rely more on the `settings` modules instead of referencing environment variables directly. These settings would all be backed by environment variables but supply defaults. Then we can update all usages of `os.environ` to instead rely on `from django.conf import settings`. This is better for null-safety and prevents mixing the two methods across the code base as we have now.

Definitely something for a follow up ticket though :D